File: src/CSharp/216.CombinationSumIII.cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.CombinationSumIII
{
public class Solution {
    static int MaxIndex = 10;
    public void traceBack(int k, int n, int sum, int begin, List<int> list, List<IList<int>> ans){
        if (k == 0){
            if (n == sum)
                ans.Add(new List<int>(list));
            return;
        }
        for (int i = begin; i < MaxIndex; i++) {
            list.Add(i);
            traceBack(k - 1, n, sum + i, i + 1, list, ans);
            list.RemoveAt(list.Count - 1);
        }
    }
    public IList<IList<int>> CombinationSum3(int k, int n) {
        List<IList<int>> ans = new List<IList<int>>();
        traceBack(k, n, 0, 1, new List<int>(), ans);
        return ans;
    }
}
}

File: src/Java/216.CombinationSumIII.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Random;

class Solution {

    public final int MAX_INDEX = 10;

    public void traceBack(int k, int n, int sum, int begin, LinkedList<Integer> list, List<List<Integer>> ans){
        if (k == 0){
            if (n == sum)
                ans.add(new ArrayList<>(list));
            return;
        }
        for (int i = begin; i < MAX_INDEX; i++) {
            list.add(i);
            traceBack(k - 1, n, sum + i, i + 1, list, ans);
            list.removeLast();
        }
    }

    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ans = new ArrayList<>();
        traceBack(k, n, 0, 1, new LinkedList<>(), ans);
        return ans;
    }
}

